/**:
  ros__parameters:
    ############ system configuration ############
    # Dataset directory for odometry
    input_dir: "${VTRDATA}/utias_20201104/nov4_b"
    # Result directory for odometry (including graph and odometry result)
    output_dir: "${VTRTEMP}/testing/stereo"
    # The sensor frame from the urdf (also defines the rig name)
    sensor_frame: front_xb3
    # The control frame from the urdf
    control_frame: base_link
    # Start and stop index
    start_index: 0 # TODO: index starts from 1 instead of 0?
    stop_index: 600

    ############ tactic configuration ############
    tactic:
      default_loc_cov: [1.0, 0.5, 0.5, 0.25, 0.25, 0.5]
      chain:
        min_cusp_distance: 1.5
        angle_weight: 7.0
        search_depth: 20
        search_back_depth: 10
        distance_warning: 3.0

    ############ pipeline configuration ############
    pipeline:
      type: stereo
      preprocessing: ["extraction", "triangulation"]
      odometry: ["recall", "matcher", "ransac", "steam", "vertex_test"]
      bundle_adjustment: ["recall", "steam"]
      localization:
        [
          "recall",
          "sub_map_extraction",
          "tod_recognition",
          "experience_triage",
          "migration",
          "matcher",
          "ransac",
          "steam",
        ]

    preprocessing:
      # conversion+extraction module
      extraction:
        type: conversion_extraction

        # Conversion-Extraction module
        # conversions: []
        # conversions: ["RGB_TO_GRAYSCALE"]
        conversions: ["RGB_TO_GRAYSCALE", "RGB_TO_COLOR_CONSTANT"]
        color_constant:
          weights: [0.43]

        extractor:
          type: ASRL_GPU_SURF # ASRL_GPU_SURF, OPENCV_ORB
          # channels: ["grayscale"]
          channels: ["grayscale", "cc_0.430000"]
          visualize_raw_features: false

          # converter/extraction/extractor/patchSize: 64
          # ORB extraction module
          orb:
            num_detector_features: 10000
            num_binned_features: 2000
            fastThreshold: 5
            x_bins: 8
            y_bins: 6
            edgeThreshold: 16
            matcher:
              stereo_descriptor_match_thresh: 0.3
              stereo_y_tolerance: 2.0
              scale_x_tolerance_by_y: false

          # SURF extraction module
          surf:
            upright_flag: true
            threshold: .000001
            nOctaves: 4
            nIntervals: 4
            initialScale: 1.5
            edgeScale: 1.5
            l1: 2.0 # 3.f/1.5f
            l2: 3.333333 # 5.f/1.5f
            l3: 2.0 # 3.f/1.5f
            l4: 0.666667 # 1.f/1.5f
            initialStep: 1
            targetFeatures: 1000
            detector_threads_x: 16
            detector_threads_y: 16
            regions_horizontal: 16
            regions_vertical: 16
            regions_target: 1000
            stereoDisparityMinimum: 0.0
            stereoDisparityMaximum: 64.0
            stereoCorrelationThreshold: .79
            stereoYTolerance: 0.9
            stereoScaleTolerance: .9

      # triangulation module
      triangulation:
        visualize_features: false
        visualize_stereo_features: false

    odometry:
      recall:
        type: landmark_recall
        landmark_source: live

      matcher:
        check_response: true
        tight_matching_pixel_thresh: 20
        matching_pixel_thresh: 400
        tight_matching_x_sigma: 0.6
        tight_matching_y_sigma: 0.6
        tight_matching_theta_sigma: 0.6
        max_point_depth: 500.0
        descriptor_thresh: 0.1
        parallel_threads: 8
        prediction_method: se3
        visualize_feature_matches: false

      ransac:
        iterations: 2000
        threshold: 5.0
        early_stop_min_inliers: 200
        min_inliers: 15
        visualize_ransac_inliers: true

      steam: # Keyframe optimisation
        iterations: 5
        pose_prior_enable: false
        ang_vel_std_dev_z: 0.2
        verbose: true
        use_T_q_m_prior: false
        # solver_type: "LevenburgMarquardt"
        depth_prior_enable: false
        trajectory_smoothing: true
        lin_acc_std_dev_x: 10.0
        lin_acc_std_dev_y: 10.0
        lin_acc_std_dev_z: 10.0
        ang_acc_std_dev_x: 1.0
        ang_acc_std_dev_y: 1.0
        ang_acc_std_dev_z: 1.0
        velocity_prior: false
        lin_vel_mean_x: 4.0
        lin_vel_mean_y: 0.0
        lin_vel_mean_z: 0.0
        ang_vel_mean_x: 0.0
        ang_vel_mean_y: 0.0
        ang_vel_mean_z: 0.0
        lin_vel_std_dev_x: 8.0
        lin_vel_std_dev_y: 3.0
        lin_vel_std_dev_z: 0.5
        ang_vel_std_dev_x: 0.5
        ang_vel_std_dev_y: 0.5
        perform_planarity_check: false
        plane_distance: 20.0
        min_point_depth: 0.0
        max_point_depth: 200.0

      vertex_test:
        match_threshold_min_count: 100
        match_threshold_fail_count: 15 # todo: better results with 25 I think

    bundle_adjustment:
      recall:
        type: stereo_windowed_recall
        window_size: 5

      steam:
        trajectory_smoothing: true
        velocity_prior: false
        verbose: false
        use_T_q_m_prior: false
        # solver_type: "LevenburgMarquardt"
        depth_prior_enable: true
        depth_prior_weight: 1000000000.0
        max_point_depth: 800.0
        min_point_depth: 1.0
        perform_planarity_check: false
        plane_distance: 20.0

    localization:
      recall:
        type: landmark_recall
        landmark_source: live

      migration:
        type: landmark_migration

      matcher:
        descriptor_thresh_cpu: 0.115
        target_match_count: 200
        min_match_count: 20
        tight_matching_pixel_thresh: 50
        matching_pixel_thresh: 200
        min_response_ratio: 0.2
        time_allowance: 2000.0
        visualize: true
        min_track_length: 1
        max_depth_diff: 5.0
        max_landmark_depth: 200.0
        screen_matched_landmarks: true

      steam:
        iterations: 15
        pose_prior_enable: true
        verbose: true
        use_T_q_m_prior: true
        # solver_type: "LevenburgMarquardt"
        depth_prior_enable: false
        trajectory_smoothing: false
        velocity_prior: false
